walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 
 package com.google.sps;
 
+import static com.google.sps.TimeRange.END_OF_DAY;
+import static com.google.sps.TimeRange.START_OF_DAY;
+import static com.google.sps.TimeRange.fromStartDuration;
+import static com.google.sps.TimeRange.fromStartEnd;
+
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> busyRanges = new LinkedList<>();
+
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        addToRanges(busyRanges, event.getWhen());
+      }
+    }
+
+    return extractFreeSlots(busyRanges).stream()
+        .filter(timeRange -> timeRange.duration() >= request.getDuration())
+        .collect(Collectors.toList());
+
+  }
+
+  private List<TimeRange> extractFreeSlots(List<TimeRange> busyRanges) {
+    List<TimeRange> freeRanges = new LinkedList<>();
+
+    int startTime = START_OF_DAY;
+
+    for (TimeRange busyRange : busyRanges) {
+
+      TimeRange freeRange = fromStartEnd(startTime, busyRange.start(), false);
+
+      if (freeRange.duration() > 0) {
+        freeRanges.add(freeRange);
+      }
+
+      startTime = busyRange.end();
+    }
+
+    freeRanges.add(fromStartEnd(startTime, END_OF_DAY, true));
+
+    return freeRanges;
+  }
+
+  private void addToRanges(List<TimeRange> ranges, TimeRange newRange) {
+    if (ranges.isEmpty()) {
+      ranges.add(newRange);
+      return;
+    }
+
+    int nextIndex = 1;
+
+    while (nextIndex < ranges.size()) {
+      if (ranges.get(nextIndex).start() >= newRange.start()) {
+        break;
+      }
+
+      nextIndex++;
+    }
+
+    ranges.add(nextIndex, newRange);
+
+    int currIndex = nextIndex - 1;
+
+    if (ranges.get(currIndex).overlaps(newRange)) {
+      mergeRanges(ranges, currIndex);
+    } else {
+      mergeRanges(ranges, nextIndex);
+    }
+  }
+
+  private void mergeRanges(List<TimeRange> ranges, int startIndex) {
```

Comment:
This function seems to be the main part of the algorithm here, and it's pretty complicated by itself. Can you add unit tests just for this function?